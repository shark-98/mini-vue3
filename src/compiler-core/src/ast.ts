export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
  ROOT
}
export const enum TagType {
  START,
  END,
}

// parse
export type contentType = string
export type contextType = {
  source: contentType
}
export interface nodeType {
  type: NodeTypes,
}
export interface interpolationType extends nodeType {
  content: nodeType & {
    content: contentType
  }
}
export interface elementType extends nodeType {
  tag: string,
  children: Array<elementType | interpolationType | textType>
}
export interface textType extends nodeType {
  content: contentType
}
export type childrenItemType = interpolationType | elementType | textType
export type childrenType = Array<childrenItemType>
export type rootType = { type: NodeTypes.ROOT, children: childrenType, codegenNode?: childrenItemType, helpers?: any[] }
export type allChildrenType = rootType | childrenItemType

// transform
export type nodeTransformsType = Function[]
export type transformOptions = {
  nodeTransforms?: nodeTransformsType
}
export type transformContext = {
  root: rootType,
  helpers: Map<string, number>,
  helper(key: any): void
} & transformOptions

// generate
export type codegenContextType = {
  code: string,
  push: (source: string) => void
  helper(key: any): string
}

export const openDelimiter = '{{'
export const closeDelimiter = '}}'
export const startTag = '<'
export const closeTag = '>'
export const endTag = '</'
