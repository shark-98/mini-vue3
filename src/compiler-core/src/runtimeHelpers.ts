export const TO_DISPLAY_STRING = Symbol("toDisplayString");

export const helperMapName: { [key: string]: string } = {
  [TO_DISPLAY_STRING]: "toDisplayString",
};
