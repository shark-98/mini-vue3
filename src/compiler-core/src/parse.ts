import { ExpandRecursively } from "../../types"
import { closeDelimiter, NodeTypes, openDelimiter } from "./ast"

type contentType = string
type contextType = {
  source: contentType
}
type childrenItemType = {
  type: NodeTypes,
  content: {
    type: NodeTypes,
    content: contentType
  }
} | undefined
type childrenType = ExpandRecursively<childrenItemType[]>
type rootType = { children: childrenType }

export const baseParse = (content: contentType): rootType => {
  const context: contextType = createParserContext(content);
  return createRoot(parseChildren(context));
}

const parseChildren = (context: contextType): childrenType => {
  const nodes: childrenType = [];

  let node;
  if (context.source.startsWith(openDelimiter)) {
    node = parseInterpolation(context);
  }

  nodes.push(node);

  return nodes;
}
const parseInterpolation = (context: contextType): childrenItemType => {
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );

  advanceBy(context, openDelimiter.length);

  const rawContentLength = closeIndex - openDelimiter.length;

  const rawContent = context.source.slice(0, rawContentLength);
  const content = rawContent.trim()

  advanceBy(context, rawContentLength + closeDelimiter.length);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content,
    },
  }
}

const createRoot = (children: childrenType): rootType => {
  return {
    children
  };
}
const createParserContext = (content: contentType): contextType => {
  return {
    source: content
  };
}
const advanceBy = (context: contextType, length: number) => {
  context.source = context.source.slice(length);
}
