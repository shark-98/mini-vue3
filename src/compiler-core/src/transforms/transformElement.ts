import { allChildrenType, createVNodeCall, elementType, NodeTypes, transformContext } from '../ast';

export const transformElement = (node: allChildrenType, context: transformContext) => {
  const nodeVal = node as elementType
  if (nodeVal.type === NodeTypes.ELEMENT) {
    return () => {
      // tag
      const vnodeTag = `'${nodeVal.tag}'`;

      // props
      let vnodeProps;

      // children
      const children = nodeVal.children;
      let vnodeChildren = children[0];

      nodeVal.codegenNode = createVNodeCall(
        context,
        vnodeTag,
        vnodeProps,
        vnodeChildren
      );
    };
  }
}
