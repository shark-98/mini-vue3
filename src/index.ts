// mini-vue 出口
import { baseCompile } from './compiler-core/src';
export * from './runtime-dom/index'
import * as runtimeDom from "./runtime-dom";
import { registerRuntimeCompiler } from "./runtime-dom";

function compileToFunction(template: string) {
  const { code } = baseCompile(template);
  const render = new Function("Vue", code)(runtimeDom);
  return render;
}

registerRuntimeCompiler(compileToFunction);
