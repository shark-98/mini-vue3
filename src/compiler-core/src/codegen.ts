import { childrenItemType, codegenContextType, interpolationType, NodeTypes, rootType, textType } from "./ast";
import { helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers";

export const codegen = (root: rootType) => {
  const context = createCodegenContext()
  const { push } = context;

  genFunctionPreamble(root, context);


  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const functionArgs = args.join(', ')
  push(`function ${functionName}(${functionArgs}) {`)

  push('return ')

  genNode(root.codegenNode, context);

  push('}')

  return {
    code: context.code
  }
};

function createCodegenContext(): codegenContextType {
  const context: codegenContextType = {
    code: '',
    push(source: string) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`;
    },
  }

  return context;
}

function genNode(node: childrenItemType | undefined, context: codegenContextType) {
  if (!node) {
    return
  }

  switch (node.type) {
    case NodeTypes.TEXT:
      genText((node as textType), context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation((node as interpolationType), context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression((node as interpolationType), context);
      break;

    default:
      break;
  }


}

function genText(node: textType, context: codegenContextType) {
  const { push } = context;
  const val = `'${node.content}'`
  push(val);
}

function genInterpolation(node: interpolationType, context: codegenContextType) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function genExpression(node: interpolationType, context: codegenContextType) {
  const { push } = context;
  push(`${node.content}`);
}

function genFunctionPreamble(ast: rootType, context: codegenContextType) {
  const { push } = context;
  const VueBinging = "Vue";
  const aliasHelper = (s: string) => `${helperMapName[s]}:_${helperMapName[s]}`;
  console.log(ast);
  
  if (ast.helpers!.length > 0) {
    push(
      `const { ${ast.helpers!.map(aliasHelper).join(", ")} } = ${VueBinging}`
    );
  }
  push("\n");
  push("return ");
}

