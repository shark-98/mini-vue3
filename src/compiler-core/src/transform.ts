import { allChildrenType, NodeTypes, rootType, transformContext, transformOptions } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export const transform = (root: rootType, options: transformOptions = {}) => {
  const context: transformContext = createTransformContext(root, options)
  traverseNode(root, context);
  createCodegenNode(root)

  root.helpers = [...context.helpers.keys()];
};

function createTransformContext(root: rootType, options: transformOptions): transformContext {
  const context: transformContext = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1);
    },
  };

  return context;
}

function traverseNode(node: allChildrenType, context: transformContext) {
  const nodeTransforms = context.nodeTransforms
  if ((Array.isArray(nodeTransforms) && nodeTransforms.length)) {
    for (let i = 0; i < nodeTransforms.length; i++) {
      const transform = nodeTransforms[i];
      transform(node)
    }
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;

    default:
      break;
  }
}

const traverseChildren = (node: allChildrenType, context: transformContext) => {
  if (!('children' in node)) {
    return
  }

  const children = node.children
  for (let i = 0; i < children.length; i++) {
    const node = children[i];
    traverseNode(node, context);
  }
}

function createCodegenNode(root: rootType) {
  root.codegenNode = root.children[0]
}

