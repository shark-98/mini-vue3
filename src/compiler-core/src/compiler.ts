import { codegen } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";

export const baseCompile = (template: string) => {

  const ast = baseParse(template)

  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  })

  return codegen(ast);
}
