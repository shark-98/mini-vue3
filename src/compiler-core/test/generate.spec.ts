import { codegen } from '../src/codegen';
import { baseParse } from '../src/parse';
import { transform } from '../src/transform';
import { transformExpression } from '../src/transforms/transformExpression';

describe('generate', () => {
  it('string', () => {
    const ast = baseParse('hi')

    transform(ast)

    const { code } = codegen(ast);

    expect(code).toMatchSnapshot()
  })

  it('interpolation', () => {
    const ast = baseParse('{{message}}')

    transform(ast, {
      nodeTransforms: [transformExpression],
    })

    const { code } = codegen(ast);

    expect(code).toMatchSnapshot()
  })
})
