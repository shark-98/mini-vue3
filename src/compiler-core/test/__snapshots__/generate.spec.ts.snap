// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate interpolation 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`generate string 1`] = `
"
return function render(_ctx, _cache) {return 'hi'}"
`;
